### ISSUE TITLE
Make assertions in autograd depending on calculator sample names

### ISSUE THREAD

**Issue by bananenpampe** [Luthaf/rascaline-torch#3]
```python
calc_2 = rascaline_torch.Calculator(rascaline.AtomicComposition(per_structure=True))

# or: calc_2 = rascaline_torch.Calculator(rascaline.SphericalExpansionByPair(**hypers_sr))

frames_2 = [rascaline_torch.as_torch_system(frame) for frame in frames_water]
composition_feat = calc_2(frames_2)

```

will lead to an assertion error in:

https://github.com/Luthaf/rascaline-torch/blob/0bff9148f99524fc41c14a3023694946523edc01/src/rascaline_torch/cxx/autograd.cpp#L352-L354

If we could make the rust BaseCalculator function samples_names part of the python/c++ api we could get the sample names in this assertion and make the assertions depending on it.

**Comment by Luthaf** [comment]
I don't understand what you mean. The code after these assert are assuming a specific set of samples, so we assert that the samples are the right ones. If we want to support other kinds of calculators/other set of samples, we need to change the code below, so there would be a `if samples_names == ["structure", "center"] {...} else if sample_names == [...]`

**Comment by bananenpampe** [comment]
yes, that is what I meant, but how do you get that info from the calculatore whether it is a structure-wise, pair-wise or atomic environment-wise calculator ?

**Comment by Luthaf** [comment]
Right, I checked and we don't need these assertions anymore in the new version, they are no longer necessary now that we have equistore-torch.

Although the new code would still likely fail if one tries to use backward with `AtomicComposition`, since it does not support any gradients. I'll have to think about that.
